core/src/main/scala/akka/kafka/internal/TransactionalSource.scala

```diff
@@ -6,24 +6,73 @@
 package akka.kafka.internal
 import java.util.Locale
 
+import akka.Done
+import akka.actor.ActorRef
+import akka.actor.Status.Failure
 import akka.annotation.InternalApi
 import akka.kafka.ConsumerMessage.TransactionalMessage
+import akka.kafka.internal.KafkaConsumerActor.Internal.Revoked
 import akka.kafka.scaladsl.Consumer.Control
-import akka.kafka.{ConsumerSettings, Subscription}
+import akka.kafka.{ConsumerMessage, ConsumerSettings, Subscription}
 import akka.stream.SourceShape
 import akka.stream.stage.GraphStageLogic
-import org.apache.kafka.clients.consumer.ConsumerConfig
+import akka.util.Timeout
+import org.apache.kafka.clients.consumer.{ConsumerConfig, ConsumerRecord, OffsetAndMetadata}
+import org.apache.kafka.common.TopicPartition
 import org.apache.kafka.common.requests.IsolationLevel
 
+import scala.collection.immutable.Iterable
+import scala.concurrent.duration.{FiniteDuration, _}
+import scala.concurrent.{Await, ExecutionContext, Future}
+
 /** Internal API */
 @InternalApi
 private[kafka] final class TransactionalSource[K, V](consumerSettings: ConsumerSettings[K, V],
-                                                     subscription: Subscription)
+                                                     subscription: Subscription,
+                                                     commitTimeout: FiniteDuration)
     extends KafkaSourceStage[K, V, TransactionalMessage[K, V]](
       s"TransactionalSource ${subscription.renderStageAttribute}"
     ) {
   require(consumerSettings.properties(ConsumerConfig.GROUP_ID_CONFIG).nonEmpty, "You must define a Consumer group.id.")
 
+  type Offset = Long
+
+  private trait InFlightRecords {
+    // Assumes that offsets per topic partition are added in the increasing order
```

Comment:
What makes this assumption hold? Can you explain it in the comment, please.

Reply:
Added:
```
    // The assumption is true for a Kafka consumer that guarantees that elements are emitted
    // per partition in offset-increasing order.```